bugfix: implement sum-preserving mode for Multinomial (#2471)

tsdlovell opened the pull request.

tsdlovell committed:
bugfix: implement sum-preserving mode for Multinomial (#2471)

tsdlovell commented:
didn't realize n could be an array

junpenglao commented:
Also, the current change breaks the following (we should actually also add this to the test):
```python
p2 = np.asarray([[.25,.25,.25,.25], [.25,.25,.25,.25]])
with pm.Model() as model:
    test7=pm.Multinomial('test7', n=1, p=p2, shape=(10, 2, 4))
```

tsdlovell commented:
changed `ifelse` to `switch`

I'm trying to think through why there is only [special logic for setting n](https://github.com/pymc-devs/pymc3/blob/0a2238889d9f29d2fb13ca9fd7bc66dda8dc2a51/pymc3/distributions/multivariate.py#L490-L491) when `len(shape)==2`.  Shouldn't it be doing that whenever `len(shape)>1`?

tsdlovell commented:
@junpenglao , added the test and implemented something that passes it.  Possibly `self.n` could be used but I don't know how to enumerate over the values inside of the tensor.

junpenglao commented on pymc3/distributions/multivariate.py:
```diff
@@ -486,6 +486,15 @@ def __init__(self, n, p, *args, **kwargs):
+                # this occurs when n is a scalar Python int or float
+                self._n *= np.ones(m, dtype='int')
+
         if len(self.shape) == 2:
```
I think you are right, here should be `len(self.shape) > 1`

junpenglao commented:
> Possibly `self.n` could be used but I don't know how to enumerate over the values inside of the tensor.

Maybe it is better to compute the mode before casting the `n` to tensor. Then you wont have the problem of looping over a tensor.

junpenglao commented:
The shape in Multinominal is always a bit of a mess. We should refactor it into something more restrict (e.g., only allow p being 1d or 2d, and the shape being 2d or 3d), but that's for another PR. For now, @tsdlovell could you please rewrite `calc_mode` into a python function and compute the mode *before* casting `n` and `p` to tensor?

In https://github.com/pymc-devs/pymc3/pull/2432#issue-244036637 there is a list of scenario should pass.

junpenglao commented:
@tsdlovell Actually, it might be better to keep `calc_mode` as a theano function as `n` and `p` could be tensors. Please ignore my previous two comments. If you can check the list of scenarios in https://github.com/pymc-devs/pymc3/pull/2432#issue-244036637 that would be great!

junpenglao commented on pymc3/distributions/multivariate.py:
```diff
@@ -486,21 +486,43 @@ def __init__(self, n, p, *args, **kwargs):
+                tt.inc_subtensor(raw_mode1d[to_increment], 1)
+            )
+
+        try:
```
why not just use `self.n` which is a tensor after line 500, so you don't need to check all these different cases?

tsdlovell replied:
IIRC, I can't use a tensor to slice, so I actually need a non-tensor.  This is a problem when the n passed in to the constructor ***is*** a tensor (as in [test_multinomial](https://github.com/tsdlovell/pymc3/blob/82dabb5d90646d7176c12ec10c18a64c99c95551/pymc3/tests/test_distributions.py#L736)), which is why I had to add the except on line 524.

junpenglao replied:
Oh I see. Let me test it locally. The problem is there are use cases where `n` is a tensor (e.g., minibatch), so this approach might not work.

junpenglao commented on pymc3/distributions/multivariate.py:
```diff
@@ -486,21 +486,43 @@ def __init__(self, n, p, *args, **kwargs):
 
         self.mean = self.n * self.p
-        self.mode = tt.cast(tround(self.mean), 'int32')
+        self._raw_mode = tt.cast(tround(self.mean), 'int32')
```
Something like this seems to work - could you try to see if it pass the tests?
```python
        rawmode = tt.cast(tround(self.mean), 'int32')
        diff = self.n - tt.sum(rawmode, axis=-1, keepdims=True)
        diffindx = diff.ravel().nonzero()
        newval = rawmode[diffindx, 0] + diff[diffindx].ravel()
        self.mode = tt.set_subtensor(rawmode[diffindx, 0], newval)
```

tsdlovell replied:
@junpenglao , it seems to work for some shapes but not all.  I think the `ravel` might not be the right thing to do, but I'm still playing with it.

tsdlovell replied:
I think an argsort needs to occur: you have to figure out which values need 1 added to them and that depends on which are furthest away from the mean.  Even if you want to just do the first n[i] for each row[i], the number could vary per row, there needs to be a way to generate rows with varying numbers of 1s.  I don't know `tensorA[tensorB, 0]` will do that.

tsdlovell replied:
I think something like this will work if the only criteria is that the sum is preserved
```
mean = n * p
rawmode = tt.cast(tt.round(mean), 'int32')
diff = n - tt.sum(rawmode, axis=-1, keepdims=True)
(n_rows, n_cols) = shape[-2:]
lst = list(range(n_cols) for _ in range(n_rows))
bool_arr = tt.as_tensor_variable(lst) < diff
mode = tt.inc_subtensor(rawmode[bool_arr.nonzero()], 1)
```

I'll keep working on how to find the "proper" indices to increment

junpenglao replied:
> you have to figure out which values need 1 added to them and that depends on which are furthest away from the mean.

Yeah I thought about that as well, there will be a bias anyway - especially in a case like `p=[.2, .2]` and `n=1`. So as long as the sum is preserved it is good enough.

tsdlovell committed:
bugfix: handle ndim > 1 case

tsdlovell committed:
TST: uncomment new passing test

dlovell commented:
@junpenglao , I think this does the trick.  Not sure why some of the jobs fail, it doesn't seem to be reporting the actual tests.  For instance: https://travis-ci.org/pymc-devs/pymc3/jobs/276355785

dlovell commented:
btw, I've added the tests from #2432.  All previously passing cases pass and one previously failing case passes.

junpenglao commented:
Great job @tsdlovell! Thanks a lot!
